astropy/utils/masked/core.py

```diff
@@ -687,7 +707,12 @@ def _combine_masks(self, masks, out=None, where=True, copy=True):
             Whether to ensure a copy is made. Only relevant if a single
             input mask is not `None`, and ``out`` is not given.
         """
-        masks = [m for m in masks if m is not None and m is not False]
+        # Simplify masks, by removing empty ones and combining possible fields.
+        masks = [
+            m if m.dtype.names is None else self._combine_fields(m)
+            for m in masks
+            if m is not None and m is not False
+        ]
```

Comment:
I'm honestly surprised how readable I find such a complex list comprehension, but I'm tempted to think it might still be less readable than a plain loop (I still kept the ternary).

Maybe performance is at stake here ? If I recall correctly comprehensions have a measurable overhead up to Python 3.11, but maybe `list.append` is the real bottleneck here ?
```suggestion
        masks = []
        for m in masks:
            if m is None or m is False:
                continue
            masks.append(m if m.dtype.names is None else self._combine_fields(m))
```

Reply:
I think I'll leave this as is - black luckily made the list comprehension fairly readable.